skyrl-train/skyrl_train/generators/skyrl_gym_generator.py

```diff
@@ -177,6 +177,24 @@ async def agent_loop(
             output_ids = engine_output["response_ids"][0]
             stop_reason = engine_output["stop_reasons"][0]
 
+            # append eos token if needed. Only applicable when sampling_params.stop is not None.
+            # Note this does not affect 3.a because the chat template adds eos_token to the end.
+            stop_strs = None
+            if sampling_params is None:
+                stop_strs = self.generator_cfg.sampling_params.get("stop", None)
+            else:
+                stop_strs = sampling_params.get("stop", None)
+            if (
+                stop_strs is not None
+                and self.generator_cfg.append_eos_token_after_stop_str_in_multi_turn
+                and (retokenize_chat_history or self.use_conversation_multi_turn)
+            ):
+                for stop_str in stop_strs:
+                    if output.endswith(stop_str) and output_ids[-1] != self.tokenizer.eos_token_id:
+                        # Append EOS token to output to match chat template termination.
+                        output_ids.append(self.tokenizer.eos_token_id)
+                        break
```

Comment:
You probably don't want to run this for every stop string, since a corner case would be when a stop str is a substring of another stop str. 

You could do:

```suggestion
                if output.endswith(tuple(stop_strs)) and output_ids[-1] != self.tokenizer.eos_token_id:
                    # Append EOS token to output to match chat template termination.
                    output_ids.append(self.tokenizer.eos_token_id)
                    break
```

Reply:
ohh that's why I added a `break` to the for loop. But yea yours is more concise, will change to this

Reply:
Changed.